lib/sinatra/base.rb

```diff
@@ -996,17 +996,15 @@ def forward
     # Accepts an optional block to call after each filter is applied.
     def filter!(type, base = settings)
       filter! type, base.superclass if base.superclass.respond_to?(:filters)
-      base.filters[type].each do |args|
-        result = process_route(*args)
-        yield result if block_given?
-      end
+      base.filters[type].each { |args| process_route(*args) }
+      @pinned_response = !response['Content-Type'].nil?
```

Comment:
Why do we need to pin the response for both before/after filters?

Reply:
Good point, I don't think we do.

Reply:
This is an excellent solution, I'll push an update now.